proposals/NNNN-outofproc-compiler-api-architecture.md

```diff
@@ -0,0 +1,408 @@
+<!-- {% raw %} -->
+
+# Out-of-process compiler api architecture
+
+* Proposal: [NNNN](NNNN-outofproc-compiler-api-architecture.md)
+* Author(s): [Cooper Partin](https://github.com/coopp)
+* Sponsor: [Cooper Partin](https://github.com/coopp)
+* Status: **Under Consideration**
+* Impacted Project(s): (Clang)
+
+* Issues:
+
+## Introduction
+
+An effort is underway to bring compilation support into clang for HLSL based
+shaders.  A C style api will be built to enable applications to compile
+a shader from their own processes in addition to being able to launching the
+clang process.
+
+This document is to propose a more detailed architecture for the approved 
+proposal: [0005](0005-inproc-outofproc-compiler-api-support.md).
+
+The c api [0008](0008-c-interface-compiler-library.md) for compiling a shader
+will be implemented using this out of process architecture.
+
+## Proposed solution
+
+The architecture for an out of process design will take advantage of systems
+that have multiple processors, or multiple-core processors. A process pool will
+be created that creates a separate compiler process for each available
+processor. For example, if the system has four processors, then four compiler
+processes are created.
+
+The process pool is owned by a singleton object and shared between each api
+instance created in the calling process.
+
+A set of low-level apis will be designed that will drive compilation work
+across multiple processes.
+
+> What do these apis look like and how are they called?
+
+Having a separate low-level api for this design enables other tooling clients
+to build their own high-level systems and implement their own policies.
+
+The high-level HLSL compiler c api implementation will be built on top of the
+low-level c apis and implement any required synchronous and asynchronous api
+policy behaviors. The high-level c api call will be blocked either waiting for
+a worker process to become available or an already dispatched work to be
+completed. This behavior depends on the type of call used
+(synchronous/asynchronous).
+
+Communication with the process pool occurs in-process to the application
```

Comment:
> application

It might be helpful to define some nomenclature for this.  I think "calling process" has been used previously to mean the same thing as "application" does here?  Or maybe not and these are different things?